Code/Framework/AzFramework/AzFramework/Visibility/OctreeSystemComponent.cpp

```diff
@@ -185,6 +185,28 @@ namespace AzFramework
         }
     }
 
+    void OctreeNode::Enumerate(const AZ::Frustum& includeFrustum, const AZ::Frustum& excludeFrustum, const IVisibilityScene::EnumerateCallback& callback) const
```

Comment:
Suggestion: And one _more_ thought to throw in the mix, maybe the API should take in a lambda for custom intersections, instead of specifically going for includeFrustum / excludeFrustum. Not sure how that would affect performance or if the benefits of the added flexibility would actually help any other use cases though.

The thought path that led me down this though was "what if somebody needs Overlaps(include) && !Overlaps(exclude)", since it would be a bummer to try and start explaining small API variations like that.

Reply:
As to the first point, while that is a common pattern, I can't think of a case for wanting that with shape intersections, at least that hasn't come up yet. I'd expect adding a lambda + conditional return would slow down code that needs to be exceptionally fast.

For the second point, I can see maybe wanting this eventually, but you again have the performance issue to deal with - the function can't optimize as well when it doesn't know what the passed in lambda is going to do (maybe a templated lambda would solve the performance issue?) Since this is the first case that's a _little_ bit different than the other cases I don't think it's worth trying to completely generalize quite yet. I could maybe see a case for Karl's idea of a "Frustum Shell" type to decrease the little bit of code duplication but I'm not sure that's worth it either, it wouldn't really reduce the amount of total code required.